Alert Plugins Conditions

mikery opened the pull request:
Hi,

I need to manage Plugin Alerts from Terraform so took a first pass at implementing #47. If this approach is acceptable and feasible to merge I'll update the PR with docs.

ctrombley commented on newrelic/resource_newrelic_alert_plugins_condition_test.go:
```diff
@@ -0,0 +1,513 @@
+		policyID := ids[0]
+		id := ids[1]
+
+		found, err := client.GetAlertCondition(policyID, id)
```
This needs to call `client.GetAlertPluginsCondition`.

ctrombley commented on newrelic/resource_newrelic_alert_plugins_condition_test.go:
```diff
@@ -0,0 +1,513 @@
+		policyID := ids[0]
+		id := ids[1]
+
+		_, err = client.GetAlertCondition(policyID, id)
```
This needs to call `client.GetAlertPluginsCondition`.

ctrombley commented on newrelic/resource_newrelic_alert_plugins_condition_test.go:
```diff
@@ -0,0 +1,513 @@
+					testAccCheckNewRelicAlertPluginsConditionExists("newrelic_alert_plugins_condition.foo"),
+					resource.TestCheckResourceAttr(
+						"newrelic_alert_plugins_condition.foo", "name", fmt.Sprintf("tf-test-%s", rName)),
+					resource.TestCheckResourceAttr(
```
This check for `type` should be deleted, since it is not a valid attribute for this resource.

ctrombley commented on newrelic/resource_newrelic_alert_plugins_condition_test.go:
```diff
@@ -0,0 +1,513 @@
+resource "newrelic_alert_plugins_condition" "foo" {
+  policy_id = "${newrelic_alert_policy.foo.id}"
+
+  name               = "tf-test-%[1]s"
```
This attribute should be set to `"tf-test-updated-%[1]s"` to match the assertion above.

ctrombley commented on newrelic/resource_newrelic_alert_plugins_condition_test.go:
```diff
@@ -0,0 +1,513 @@
+  policy_id = "${newrelic_alert_policy.foo.id}"
+
+  name               = "tf-test-%[1]s"
+  enabled            = false
```
This attribute should be set to `true` to match the assertion above.

ctrombley commented on newrelic/resource_newrelic_alert_plugins_condition_test.go:
```diff
@@ -0,0 +1,513 @@
+  name               = "tf-test-%[1]s"
+  enabled            = false
+  entities           = ["${data.newrelic_application.app.id}"]
+  runbook_url        = "https://foo.example.com"
```
This attribute should be set to `"https://bar.example.com"` to match the assertion above.

ctrombley commented on newrelic/resource_newrelic_alert_plugins_condition_test.go:
```diff
@@ -0,0 +1,513 @@
+
+func testAccCheckNewRelicAlertPluginsConditionDestroy(s *terraform.State) error {
+	client := testAccProvider.Meta().(*ProviderConfig).Client
+	for _, r := range s.RootModule().Resources {
```
Required at the top of this loop:

```
if r.Type != "newrelic_plugins_alert_condition" {
   continue
}
```
This is since there is also an alert policy defined in state at this point.

mikery commented:
Thanks @ctrombley for the quick response and detailed review!

> * Can we change the name of the resource definition func and associated file for consistency with the others?  `resourceNewRelicPluginsAlertCondition` lines up a bit better.  All references in the test file's HCL to the `newrelic_alert_plugins_condition` label will need to be updated to match as well.

I originally named it `NewRelicPluginsAlertCondition`, but renamed to `..AlertPluginsCondition` for consistency with the NR API and Go newrelic library. `newrelic_plugins_alert_condition` is a much clearer name, so will happily change back :-) 

> * This seems to call for a data source for Plugins, but that may be out of the scope of this PR.

I would also find this useful. Would a minimal implementation that allows looking up the plugin ID via its GUID be sufficient?
```
data "newrelic_plugin" "my_plugin" {
  guid = "com.example.my_plugin"
}

resource "foo" "bar" {
  plugin_id = "${data.newrelic_plugin.my_plugin.plugin_id}"
}
```
If that would be ok I can include it with this PR. It solves my immediate use case (no need to hardcode plugin IDs in my TF manifests).

If we'd need more functionality (e.g. access to the list of summary metrics), I'd be happy to work on it if you can scope it out in a separate issue and give some example usages you'd like to see.

> * In the meantime, I encourage you to try and get the acceptance tests running against your own account.  You can do this with `NEWRELIC_API_KEY=<api_key> NEWRELIC_LICENSE_KEY=<license_key> make testacc`.  Ideally we will have those passing once the feedback I've provided here has been implemented.

Cool, I'll try that after making the requested changes.

mikery commented:
@ctrombley I've applied the requested changes, added docs, and added a basic data source for looking up plugins. I haven't yet tried the integration tests as I'd like to set up an isolated NR account for this.

ctrombley commented:
Thanks for all your work on this PR @mikery!  I rebased on master, added a plugin component data source, and refactored the acceptance tests so that they are passing on our internal account.  I'm going to have someone else on my team do a final review and then we'll merge this in.  

We are going to stick with "newrelic_plugins_alert_conditions" for consistency with the underlying API, even though I do agree that it sounds clearer in the singular when spoken.

I am curious to know if it's possible to create a plugins alert condition without referencing an entity (plugin component), as suggested in the docs you wrote for the plugin data source.  Do you have an example of a plugin that behaves this way?

We will eventually chase this up with another PR to fill in the detailed metric data for the two new data sources.

ctrombley commented:
Resolves #47.

mikery commented:
> Thanks for all your work on this PR @mikery!
My pleasure, thanks for getting this merged in and adding the other data source!

> I am curious to know if it's possible to create a plugins alert condition without referencing an entity (plugin component), as suggested in the docs you wrote for the plugin data source. Do you have an example of a plugin that behaves this way?

That was an oversight on my part. The entities list is required. 
> 
> We will eventually chase this up with another PR to fill in the detailed metric data for the two new data sources.

Please cc/@ me when you add an issue for this. I'll try to pick it up, schedule permitting.
